File: host/detector.ts
import type { CsiFrame } from "./protocol.ts";

const HISTORY_LENGTH = 120;
const UINT32_MODULUS = 0x1_0000_0000;

function normalizedAmplitudes(samples: readonly number[]): number[] {
  const amplitudes: number[] = [];
  for (let index = 0; index < samples.length; index += 2) {
    amplitudes.push(Math.hypot(samples[index], samples[index + 1]));
  }
  const rms = Math.sqrt(
    amplitudes.reduce((sum, value) => sum + value * value, 0) / amplitudes.length,
  );
  if (rms < 1e-6) {
    return [];
  }
  return amplitudes.map((value) => value / rms);
}

function round(value: number, digits: number): number {
  return Number(value.toFixed(digits));
}

export class LinkState {
  readonly receiver: string;
  readonly calibrationSamples: number;
  baseline: number[] = [];
  calibrationCount = 0;
  calibrationDeviations: number[] = [];
  threshold = 0.055;
  score = 0;
  active = false;
  hitStreak = 0;
  clearStreak = 0;
  frames = 0;
  dropped = 0;
  rssi = 0;
  noiseFloor = 0;
  channel = 0;
  lastSequence: number | null = null;
  lastSeen = 0;
  history: number[] = [];

  constructor(receiver: string, calibrationSamples: number) {
    this.receiver = receiver;
    this.calibrationSamples = calibrationSamples;
  }

  get calibrated(): boolean {
    return this.calibrationCount >= this.calibrationSamples;
  }

  resetShape(vector: number[]): void {
    this.baseline = [...vector];
    this.calibrationCount = 1;
    this.calibrationDeviations = [];
    this.threshold = 0.055;
    this.score = 0;
    this.active = false;
    this.hitStreak = 0;
    this.clearStreak = 0;
    this.history = [];
  }

  ingest(frame: CsiFrame, now: number): boolean {
    const vector = normalizedAmplitudes(frame.samples);
    if (vector.length < 8) {
      return false;
    }

    this.frames += 1;
    this.dropped += frame.dropped;
    if (this.lastSequence !== null) {
      const expected = (this.lastSequence + 1) % UINT32_MODULUS;
      if (frame.sequence !== expected) {
        const gap = (frame.sequence - expected + UINT32_MODULUS) % UINT32_MODULUS;
        if (gap < 1_000_000) {
          this.dropped += gap;
        }
      }
    }
    this.lastSequence = frame.sequence;
    this.rssi = frame.rssi;
    this.noiseFloor = frame.noiseFloor;
    this.channel = frame.channel;
    this.lastSeen = now;

    if (vector.length !== this.baseline.length) {
      this.resetShape(vector);
      return false;
    }

    const deviation = Math.sqrt(
      vector.reduce((sum, value, index) => {
        const difference = value - this.baseline[index];
        return sum + difference * difference;
      }, 0) / vector.length,
    );

    if (!this.calibrated) {
      this.calibrationCount += 1;
      const alpha = 1 / this.calibrationCount;
      vector.forEach((value, index) => {
        this.baseline[index] += alpha * (value - this.baseline[index]);
      });
      if (this.calibrationCount > 5) {
        this.calibrationDeviations.push(deviation);
      }
      if (this.calibrated) {
        const mean =
          this.calibrationDeviations.reduce((sum, value) => sum + value, 0) /
          this.calibrationDeviations.length;
        const spread =
          this.calibrationDeviations.length > 1
            ? Math.sqrt(
                this.calibrationDeviations.reduce((sum, value) => sum + (value - mean) ** 2, 0) /
                  this.calibrationDeviations.length,
              )
            : 0;
        this.threshold = Math.max(0.055, mean + 5 * spread);
      }
      this.score = 0;
      this.appendHistory(0);
      return false;
    }

    const rawScore = deviation / this.threshold;
    this.score = 0.7 * this.score + 0.3 * rawScore;
    this.appendHistory(Math.min(this.score, 4));

    if (this.score >= 1) {
      this.hitStreak += 1;
      this.clearStreak = 0;
      if (this.hitStreak >= 3) {
        this.active = true;
      }
    } else if (this.score < 0.55) {
      this.hitStreak = 0;
      this.clearStreak += 1;
      if (this.clearStreak >= 8) {
        this.active = false;
      }
    } else {
      this.hitStreak = Math.max(0, this.hitStreak - 1);
      this.clearStreak = 0;
    }

    if (!this.active && this.score < 0.7) {
      vector.forEach((value, index) => {
        this.baseline[index] += 0.001 * (value - this.baseline[index]);
      });
    }
    return this.active;
  }

  private appendHistory(value: number): void {
    this.history.push(value);
    if (this.history.length > HISTORY_LENGTH) {
      this.history.shift();
    }
  }
}

export interface LinkSnapshot {
  receiver: string;
  calibrated: boolean;
  calibration: number;
  active: boolean;
  score: number;
  rssi: number;
  noiseFloor: number;
  channel: number;
  frames: number;
  dropped: number;
  lastSeenSeconds: number;
  history: number[];
}

export interface DetectorSnapshot {
  state: "offline" | "calibrating" | "occupied" | "clear";
  occupied: boolean;
  score: number;
  calibration: number;
  holdRemainingSeconds: number;
  generation: number;
  links: LinkSnapshot[];
}

/** Combines independent CSI links into one held occupancy state. */
export class RoomDetector {
  readonly calibrationSamples: number;
  readonly holdSeconds: number;
  readonly links = new Map<string, LinkState>();
  lastActivity: number | null = null;
  generation = 0;

  constructor(calibrationSamples = 80, holdSeconds = 20) {
    if (calibrationSamples < 10) {
      throw new Error("calibrationSamples must be at least 10");
    }
    this.calibrationSamples = calibrationSamples;
    this.holdSeconds = holdSeconds;
  }

  ingest(frame: CsiFrame, now = performance.now() / 1000): void {
    let link = this.links.get(frame.receiver);
    if (!link) {
      link = new LinkState(frame.receiver, this.calibrationSamples);
      this.links.set(frame.receiver, link);
    }
    if (link.ingest(frame, now)) {
      this.lastActivity = now;
    }
  }

  reset(): void {
    this.links.clear();
    this.lastActivity = null;
    this.generation += 1;
  }

  snapshot(now = performance.now() / 1000): DetectorSnapshot {
    const allLinks = [...this.links.values()];
    const liveLinks = allLinks.filter((link) => now - link.lastSeen <= 2);
    const allCalibrated = liveLinks.length > 0 && liveLinks.every((link) => link.calibrated);
    const occupied =
      allCalibrated && this.lastActivity !== null && now - this.lastActivity <= this.holdSeconds;

    let state: DetectorSnapshot["state"];
    if (liveLinks.length === 0) {
      state = "offline";
    } else if (!allCalibrated) {
      state = "calibrating";
    } else if (occupied) {
      state = "occupied";
    } else {
      state = "clear";
    }

    const links = allLinks
      .sort((left, right) => left.receiver.localeCompare(right.receiver))
      .map((link): LinkSnapshot => ({
        receiver: link.receiver,
        calibrated: link.calibrated,
        calibration: Math.min(1, link.calibrationCount / link.calibrationSamples),
        active: link.active,
        score: round(link.score, 3),
        rssi: link.rssi,
        noiseFloor: link.noiseFloor,
        channel: link.channel,
        frames: link.frames,
        dropped: link.dropped,
        lastSeenSeconds: round(now - link.lastSeen, 2),
        history: link.history.map((value) => round(value, 3)),
      }));

    const score = Math.max(0, ...liveLinks.map((link) => link.score));
    const calibration =
      liveLinks.length === 0
        ? 0
        : Math.min(...liveLinks.map((link) => link.calibrationCount / link.calibrationSamples));
    const holdRemaining =
      this.lastActivity === null ? 0 : Math.max(0, this.holdSeconds - (now - this.lastActivity));

    return {
      state,
      occupied,
      score: round(score, 3),
      calibration: round(calibration, 3),
      holdRemainingSeconds: round(holdRemaining, 1),
      generation: this.generation,
      links,
    };
  }
}

File: host/detector.py
"""Gain-resistant CSI baseline detector for the room demo."""

from __future__ import annotations

import math
import statistics
import threading
import time
from collections import deque
from dataclasses import dataclass, field

from host.protocol import CsiFrame


def _normalized_amplitudes(samples: tuple[int, ...]) -> list[float]:
    amplitudes = [
        math.hypot(samples[index], samples[index + 1])
        for index in range(0, len(samples), 2)
    ]
    rms = math.sqrt(sum(value * value for value in amplitudes) / len(amplitudes))
    if rms < 1e-6:
        return []
    return [value / rms for value in amplitudes]


@dataclass
class LinkState:
    receiver: str
    calibration_samples: int
    baseline: list[float] = field(default_factory=list)
    calibration_count: int = 0
    calibration_deviations: list[float] = field(default_factory=list)
    threshold: float = 0.055
    score: float = 0.0
    active: bool = False
    hit_streak: int = 0
    clear_streak: int = 0
    frames: int = 0
    dropped: int = 0
    rssi: int = 0
    noise_floor: int = 0
    channel: int = 0
    last_sequence: int | None = None
    last_seen: float = 0.0
    history: deque[float] = field(default_factory=lambda: deque(maxlen=120))

    @property
    def calibrated(self) -> bool:
        return self.calibration_count >= self.calibration_samples

    def reset_shape(self, vector: list[float]) -> None:
        self.baseline = vector.copy()
        self.calibration_count = 1
        self.calibration_deviations.clear()
        self.threshold = 0.055
        self.score = 0.0
        self.active = False
        self.hit_streak = 0
        self.clear_streak = 0
        self.history.clear()

    def ingest(self, frame: CsiFrame, now: float) -> bool:
        vector = _normalized_amplitudes(frame.samples)
        if len(vector) < 8:
            return False

        self.frames += 1
        self.dropped += frame.dropped
        if self.last_sequence is not None:
            expected = (self.last_sequence + 1) & 0xFFFFFFFF
            if frame.sequence != expected:
                gap = (frame.sequence - expected) & 0xFFFFFFFF
                # A small forward gap is loss. A large modular gap means the
                # receiver rebooted and restarted its local sequence counter.
                if gap < 1_000_000:
                    self.dropped += gap
        self.last_sequence = frame.sequence
        self.rssi = frame.rssi
        self.noise_floor = frame.noise_floor
        self.channel = frame.channel
        self.last_seen = now

        if len(vector) != len(self.baseline):
            self.reset_shape(vector)
            return False

        deviation = math.sqrt(
            sum((value - mean) ** 2 for value, mean in zip(vector, self.baseline))
            / len(vector)
        )

        if not self.calibrated:
            self.calibration_count += 1
            alpha = 1.0 / self.calibration_count
            for index, value in enumerate(vector):
                self.baseline[index] += alpha * (value - self.baseline[index])
            if self.calibration_count > 5:
                self.calibration_deviations.append(deviation)
            if self.calibrated:
                mean = statistics.fmean(self.calibration_deviations)
                spread = (
                    statistics.pstdev(self.calibration_deviations)
                    if len(self.calibration_deviations) > 1
                    else 0.0
                )
                self.threshold = max(0.055, mean + 5.0 * spread)
            self.score = 0.0
            self.history.append(0.0)
            return False

        raw_score = deviation / self.threshold
        self.score = 0.7 * self.score + 0.3 * raw_score
        self.history.append(min(self.score, 4.0))

        if self.score >= 1.0:
            self.hit_streak += 1
            self.clear_streak = 0
            if self.hit_streak >= 3:
                self.active = True
        elif self.score < 0.55:
            self.hit_streak = 0
            self.clear_streak += 1
            if self.clear_streak >= 8:
                self.active = False
        else:
            self.hit_streak = max(0, self.hit_streak - 1)
            self.clear_streak = 0

        if not self.active and self.score < 0.7:
            for index, value in enumerate(vector):
                self.baseline[index] += 0.001 * (value - self.baseline[index])
        return self.active


class RoomDetector:
    """Combines independent CSI links into one held occupancy state."""

    def __init__(self, calibration_samples: int = 80, hold_seconds: float = 20.0):
        if calibration_samples < 10:
            raise ValueError("calibration_samples must be at least 10")
        self.calibration_samples = calibration_samples
        self.hold_seconds = hold_seconds
        self.links: dict[str, LinkState] = {}
        self.last_activity: float | None = None
        self.generation = 0
        self._lock = threading.RLock()

    def ingest(self, frame: CsiFrame, now: float | None = None) -> None:
        instant = time.monotonic() if now is None else now
        with self._lock:
            link = self.links.get(frame.receiver)
            if link is None:
                link = LinkState(frame.receiver, self.calibration_samples)
                self.links[frame.receiver] = link
            if link.ingest(frame, instant):
                self.last_activity = instant

    def reset(self) -> None:
        with self._lock:
            self.links.clear()
            self.last_activity = None
            self.generation += 1

    def snapshot(self, now: float | None = None) -> dict[str, object]:
        instant = time.monotonic() if now is None else now
        with self._lock:
            live_links = [
                link for link in self.links.values() if instant - link.last_seen <= 2.0
            ]
            all_calibrated = bool(live_links) and all(
                link.calibrated for link in live_links
            )
            occupied = (
                all_calibrated
                and self.last_activity is not None
                and instant - self.last_activity <= self.hold_seconds
            )
            if not live_links:
                state = "offline"
            elif not all_calibrated:
                state = "calibrating"
            elif occupied:
                state = "occupied"
            else:
                state = "clear"

            links = []
            for link in sorted(self.links.values(), key=lambda item: item.receiver):
                links.append(
                    {
                        "receiver": link.receiver,
                        "calibrated": link.calibrated,
                        "calibration": min(
                            1.0, link.calibration_count / link.calibration_samples
                        ),
                        "active": link.active,
                        "score": round(link.score, 3),
                        "rssi": link.rssi,
                        "noiseFloor": link.noise_floor,
                        "channel": link.channel,
                        "frames": link.frames,
                        "dropped": link.dropped,
                        "lastSeenSeconds": round(instant - link.last_seen, 2),
                        "history": [round(value, 3) for value in link.history],
                    }
                )

            return {
                "state": state,
                "occupied": occupied,
                "score": round(max((link.score for link in live_links), default=0.0), 3),
                "calibration": round(
                    min(
                        (link.calibration_count / link.calibration_samples for link in live_links),
                        default=0.0,
                    ),
                    3,
                ),
                "holdRemainingSeconds": round(
                    max(
                        0.0,
                        self.hold_seconds
                        - (instant - self.last_activity)
                        if self.last_activity is not None
                        else 0.0,
                    ),
                    1,
                ),
                "generation": self.generation,
                "links": links,
            }
